pyteal/ast/router.py

```diff
@@ -0,0 +1,547 @@
+    """
+    CallConfigs: a "bitset" alike class for more fine-grained control over
+    `call or create` for a method about an OnComplete case.
+
+    This enumeration class allows for specifying one of the four following cases:
+    - CALL
+    - CREATE
+    - ALL
+    - NEVER
+    for a method call on one on_complete case.
+    """
+
+    NEVER = 0
+    CALL = 1
+    CREATE = 2
+    ALL = 3
+
+    def __or__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return CallConfig(self.value | other.value)
+
+    def __and__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return CallConfig(self.value & other.value)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return self.value == other.value
+
+
+CallConfig.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CallConfigs:
+    """
+    CallConfigs keep track of one method registration's CallConfigs for all OnComplete cases.
+    """
+
+    no_op: CallConfig = field(kw_only=True, default=CallConfig.CALL)
+    opt_in: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    close_out: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    clear_state: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    update_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    delete_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+
+    def is_never(self) -> bool:
+        return (
+            self.no_op == CallConfig.NEVER
+            and self.opt_in == CallConfig.NEVER
+            and self.close_out == CallConfig.NEVER
+            and self.clear_state == CallConfig.NEVER
+            and self.update_application == CallConfig.NEVER
+            and self.delete_application == CallConfig.NEVER
+        )
+
+    def oc_under_call_config(self, call_config: CallConfig) -> list[EnumInt]:
+        if not isinstance(call_config, CallConfig):
+            raise TealInputError(
+                "generate condition based on OCMethodCallConfigs should be based on OCMethodConfig"
+            )
+        config_oc_pairs: list[tuple[CallConfig, EnumInt]] = [
+            (self.no_op, OnComplete.NoOp),
+            (self.opt_in, OnComplete.OptIn),
+            (self.close_out, OnComplete.CloseOut),
+            (self.clear_state, OnComplete.ClearState),
+            (self.update_application, OnComplete.UpdateApplication),
+            (self.delete_application, OnComplete.DeleteApplication),
+        ]
+        return [
+            oc
+            for oc_config, oc in config_oc_pairs
+            if (oc_config & call_config) != CallConfig.NEVER
```

Comment:
not quite, so `&` is overriden by `__and__` method in `CallConfig`, but `and` is a logical operator, not an arithmetic operator in python.

If we want to work in the suggested direction, we want to also override `__bool__` method, while I think we might not want to have boolean property over `CallConfig`, since it is more versatile than boolean's expressibility.

Reply:
not worth the change IMO